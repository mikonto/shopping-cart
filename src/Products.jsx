import styled from "styled-components";
import { useEffect, useState } from "react";
import ProductCard from "./ProductCard";

const StyledProducts = styled.div`
  background: white;
  flex-grow: 1;
  display: grid;
  gap: 4px;
  padding: 4px;
  grid-auto-rows: minmax(250px, auto);
  grid-template-columns: repeat(4, 1fr);

  @media (max-width: 1200px) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media (max-width: 600px) {
    grid-template-columns: repeat(1, 1fr);
  }
`;

const Products = () => {
  const [products, setProducts] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    fetch("https://fakestoreapi.com/products")
      .then((response) => response.json())
      .then((response) => {
        // Limit the array to the first 9 items
        const limitedProducts = response.slice(0, 8);
        setProducts(limitedProducts);
        console.log(limitedProducts);
      })
      .catch((error) => setError(error))
      .finally(() => setLoading(false));
  }, []);

  if (error) return <p>A network error was encountered</p>;
  if (loading) return <p>Loading...</p>;

  return (
    <>
      <StyledProducts>
        {products.map((product) => (
          <ProductCard
            key={product.id}
            title={product.title}
            price={product.price}
            description={product.description}
            category={product.category}
            rating={product.rating.rate}
            ratingCount={product.rating.count}
            image={product.image}
          />
        ))}
      </StyledProducts>
    </>
  );
};

export default Products;
