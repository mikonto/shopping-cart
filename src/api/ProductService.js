const fetchProducts = () => {
    return fetch("https://fakestoreapi.com/products")
      .then((response) => response.json())
      .then((products) => products.slice(0, 8)); // limiting to 8 products
  };
  
  export { fetchProducts };
  