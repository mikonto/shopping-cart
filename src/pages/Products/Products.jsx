import styled from "styled-components";
import { useEffect, useState } from "react";
import ProductCard from "./ProductCard";
import PropTypes from "prop-types";
import { fetchProducts } from "../../api/ProductService";

const StyledProducts = styled.div`
  flex-grow: 1;
  display: grid;
  gap: 12px;
  padding: 12px;
  grid-auto-rows: minmax(250px, auto);
  grid-template-columns: repeat(4, 1fr);

  @media (max-width: 1200px) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media (max-width: 600px) {
    grid-template-columns: repeat(1, 1fr);
  }
`;

const Title = styled.h1`
  background: #f6f6f6;
  padding: 12px;
  margin: 8px;
`;

const Products = ({ addToCart }) => {
  const [products, setProducts] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    fetchProducts()
      .then((products) => setProducts(products))
      .catch((error) => setError(error))
      .finally(() => setLoading(false));
  }, []);

  if (error)
    return (
      <StyledProducts>
        <p>A network error was encountered</p>
      </StyledProducts>
    );
  if (loading)
    return (
      <StyledProducts>
        <p>Loading...</p>
      </StyledProducts>
    );

  return (
    <>
      <Title>Products</Title>
      <StyledProducts>
        {products.map((product) => (
          <ProductCard
            key={product.id}
            product={product}
            addToCart={addToCart}
          />
        ))}
      </StyledProducts>
    </>
  );
};

Products.propTypes = {
  addToCart: PropTypes.func,
};

export default Products;
