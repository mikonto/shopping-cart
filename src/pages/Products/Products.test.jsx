import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { vi } from "vitest";
import Products from "./Products";
import { fetchProducts } from "../../api/ProductService";

vi.mock("../../api/ProductService");
describe("Products component", () => {
  const mockProducts = [
    {
      id: 1,
      title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
      price: 109.95,
      description:
        "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
      category: "men's clothing",
      image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
      rating: {
        rate: 3.9,
        count: 120,
      },
    },
  ];

  it("displays loading initially", () => {
    fetchProducts.mockResolvedValue(mockProducts);
    render(<Products addToCart={() => {}} />);
    expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
  });

  it("displays an error message if the fetch fails", async () => {
    fetchProducts.mockRejectedValue(new Error("Network error"));

    render(
      <MemoryRouter>
        <Products addToCart={() => {}} />
      </MemoryRouter>
    );
    const errorMessage = await screen.findByText(
      /A network error was encountered/i
    );
    expect(errorMessage).toBeInTheDocument();
  });

  it("renders the correct title", async () => {
    fetchProducts.mockResolvedValue(mockProducts);
    render(<Products addToCart={() => {}} />);

    const title = await screen.findByRole("heading", { name: /products/i });
    expect(title).toBeInTheDocument();
  });

  it("renders ProductCard components", async () => {
    fetchProducts.mockResolvedValue(mockProducts);
    render(<Products addToCart={() => {}} />);

    for (const product of mockProducts) {
      // Assuming the title is rendered in each ProductCard
      expect(await screen.findByText(product.title)).toBeInTheDocument();
    }
  });

  it("it renders correctly", () => {
    const { container } = render(<Products addToCart={() => {}} />);
    expect(container).toMatchSnapshot();
  });
});
