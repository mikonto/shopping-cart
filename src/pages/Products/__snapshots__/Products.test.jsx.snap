// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products component > it renders correctly 1`] = `
<div>
  <div
    class="sc-iBYPxa bYcRKg"
  >
    <p>
      Loading...
    </p>
  </div>
</div>
`;
